### ISSUE TITLE
Inspect the random failures in Cypress

### ISSUE THREAD

**Issue by jonasraoni** [pkp/pkp-lib#7601]
**Describe the bug**
The Cypress tests might fail due to unexpected conditions.
- Accessing a given control while it's not ready (e.g. attempting to click in an element which is: covered by another, currently detached from the DOM, still being built by Vue, ...)
- Possibly racing conditions on the requests. Locally I have a random failure while creating the users: in the middle of the process, the current user gets logged off, and when attempting to open the "Add User" popup to create the next one, the popup is closed right away.

**Draft of ideas**
- Running in a slower environment might be helpful.
- Check if increasing [default timeouts](https://docs.cypress.io/guides/references/configuration#Timeouts) help (in case we have slow requests/operations breaking tests).
- Investigate other approaches out there, regarding knowing if the interface is built and not busy.
    - Intercept the requests and add a general check `cy.isIdle()` (if all requests are resolved).
    - Given we own the UI components, we could add a `ready`/`busy` class to the controls and implement a `cy.isUiReady()` `cy.isUiBusy()` (sounds like a bazooka to kill ants).
    - Etc.
- If there's no silver bullet, then ensure we're waiting for the right events and fallback to adding some sleeping time when needed.

**What application are you using?**
*/main

**Comment by NateWr** [comment]
This problem is much more prominent in parts of the code that use the old JS library. That's because it tends to update large chunks of the DOM at once. Where the new Vue-based UI is used, Vue is smart enough to make smaller updates, so fewer references to DOM objects get trashed.

Where this can still be an issue is in waiting for requests to finish. But in almost all cases, the UI _should_ provide progress updates. If this information isn't available, that's usually a sign of a UX problem. For example, when a form is submitted, a test shouldn't just look for the new data. It should look for the resulting notification (like "`cy.get('#my-form').contains('Saved');`) to ensure that these functions of the UI also get tested. Scripting tests in this way also tends to minimize cases where Cypress looks for data before it is available.

**Comment by jonasraoni** [comment]
This is the most common message when a build fails partially:
`Expected to find content: 'PDF Version 2' within the element: <a#cell-2-label-downloadFile-button-6200e0551b407.pkp_controllers_linkAction.pkp_linkaction_downloadFile.pkp_linkaction_icon_> but never did`

Perhaps fixing it will be enough to decrease the number of rebuilds 🤔

**Comment by NateWr** [comment]
Yeah I see that one quite a bit too. Usually what this means is that there is a race condition so that between the type that Cypress `.get()`s a DOM element and checks if it `.contains()` something, that DOM element has been removed. 

Since Cypress does everything async, a `cy.get(something).contains()` will work in two separate async callbacks, first to `.get(something)` and then to check `something.contains(...)`.

I _think_ it might fix it if you change the following:

```js
cy.get('#representations-grid [id*="downloadFile-button"').contains('PDF Version 2');
```

So that the text check happens in the same call as the `.get()`:

```js
cy.get('#representations-grid [id*="downloadFile-button"]:contains("PDF Version 2")');
```

**Comment by jonasraoni** [comment]
Looks like the modification above fixed the issue, I've added it to a random PR and I haven't seen this issue lately.
I'll try to run the tests in my machine again to see if I can fix the other random issues I had in my sluggy env =]

**Comment by jonasraoni** [comment]
I'll close this issue for now. I believe that once we update Cypress to a newer version (we're near to Cypress 10.x) most of these issues will be gone =]
